ufl/core/external_operator.py

```diff
@@ -0,0 +1,98 @@
+from ufl.utils.str import as_native_str
+from ufl.coefficient import Coefficient
+from ufl.core.ufl_type import ufl_type
+from ufl.constantvalue import as_ufl
+from ufl.log import error
+from ufl.finiteelement.mixedelement import TensorElement
+from ufl.functionspace import FunctionSpace
+from ufl.domain import default_domain
+import weakref
+
+
+@ufl_type(inherit_indices_from_operand=0, is_terminal=True, is_differential=True)
```

Comment:
A terminal should have no operands. This appears not to be the case here.

Reply:
In principle yes, but here ExternalOperator needs to be a Terminal because it is  seen as a Coefficient, it is a Coefficient that gets evaluated inside the assembly of a given form. It has operands because of its nature, that is why we end up with this duality and with this extended notion of Terminal.

Reply:
So I somehow would prefer a separation between the "here is a coefficient" and "here is a way of computing a coefficient" which I think pointwise operator provides. There seems to be a separation between what is needed for the assembly kernel _once we've computed the values_ (you just need to see a coefficient), and the object that can compute the values (needed outside?).

Perhaps the right thing to do is to have part of the compute_form_data pipeline _replace_ an `ExternalOperator` with a coefficient on the same space and let it remember what the mapping is. Then the form correctly advertises what coefficients it depends on (I think?).

Reply:
They are not only a way of computing coefficients they are also coefficients we need the value of, for instance if we think of solving a nonlinear relationship pointwisely, your pointwise operator will not only contain the information about what relation it is that you have to solve but it also represents the quantity of interest (i.e. solution of your nonlinear relations).

So when assembling, we call its evaluate method which behind the scenes acts on the pointwise operator itself and changes its value (cf. https://github.com/firedrakeproject/firedrake/blob/831edc7a6fb788ae6e54b16926b56c8f817904a6/firedrake/assemble.py#L437).

Aside from this, they will mostly be treated all over the pipeline as normal coefficients, which they are by nature.